ocaml/forkexecd/helper/vfork_helper.c

```diff
@@ -335,14 +335,44 @@ reset_signal_handlers(void)
 static void
 clear_cgroup(void)
 {
-    int fd = open("/sys/fs/cgroup/systemd/cgroup.procs", O_WRONLY|O_CLOEXEC);
-    if (fd >= 0) {
-        char string_pid[32];
-        int ignored __attribute__((unused));
-        sprintf(string_pid, "%d\n", (int) getpid());
-        ignored = write(fd, string_pid, strlen(string_pid));
+    // list of files to try, terminated by NULL
+    static const char *cgroup_files[] = {
+        "/sys/fs/cgroup/systemd/cgroup.procs",
+        "/sys/fs/cgroup/cgroup.procs",
+        NULL
+    };
+
+    char string_pid[32];
+    int last_error = 0;
+    const char *last_error_msg = NULL;
+    const char *last_fn = NULL;
+
+    snprintf(string_pid, sizeof(string_pid), "%d\n", (int) getpid());
+
+    for (const char **fn = cgroup_files; *fn != NULL; ++fn) {
+        last_fn = *fn;
+        int fd = open(*fn, O_WRONLY|O_CLOEXEC);
+        if (fd < 0) {
+            last_error = errno;
+            last_error_msg = "Error opening file %s";
+            continue;
+        }
+
+        ssize_t written = write(fd, string_pid, strlen(string_pid));
```

Comment:
`write` can in general return a short count (e.g. on network sockets). I don't know whether that could happen on sysfs or not, but it might be better to either loop here and ensure we've written out all the bytes, or use `fopen`/`fwrite` instead (which I think would do the looping internally, since it is documented to return short counts only only on error/EOF).

I can see why you might be reluctant to use `fwrite` (I used to be too, because it gave me more reliable error handling, or so I thought), because the manpages are wrong and don't tell you that it can also set `errno`. But I intend to send a patch to fix the manpages (the POSIX ones and the libc implementations actually do set errno on `fwrite`).

Reply:
No, in this case you cannot have shorter write, the kernel parses the number and use it. Well, to confirm, you need to look at the kernel and check buffers. In this case, `write` is also used here to have more control, although the default buffer of `FILE` i/o and the way buffers are flushed won't be much of a difference (but probably you would have to use also `fflush` (otherwise potentially `fwrite` will just write into memory buffer and return success depending on buffer settings, which usually for a file is full buffer).

Reply:
I'll add a comment about it